digitalfabrik/lunes-cms
# Packaging & release workflow

### Short description

Automatically build package and create releases when changes are merged into the main branch
I'm pretty happy with the workflow of the [integreat-cms](https://github.com/digitalfabrik/integreat-cms/blob/develop/.circleci/config.yml), so most of it ist just copied + pasted from there.

### Proposed changes

- Move setup configuration to setup.cfg
- Remove unused dependencies
- Pin dependency versions in setup.cfg (fixes #319)
- Rename src directory to lunes_cms (fixes #67)
- Squash migrations
- Move tests outside of lunes_cms package
- Install system dependencies before testing
- Apply black code style on whole project (fixes #266)
- Use automatic calendar versioning (CalVer) (fixes #280)
- Add CHANGELOG.md (fixes #326)
- Publish pip package to (Test-)PyPI (fixes #327)
- Automatically create GitHub release (fixes #328)
- Automatically send Mattermost release notifications
- Rename dev tools directory
- Improve dev tool utility functions
- Import test user from fixture
- Add installation dev tool
- Add pre-commit hooks
- Add runserver dev tool
- Add translation dev tools (fixes #264)
- Remove compiled translation files from git
-  Update development setup documentation
- Add shellcheck to test tools
- Remove outdated docker files
- Remove outdated systemd config file
- Improve Lunes logos
- Fix and improve documentation build
- Move sphinx directory to docs/src
- Improve documentation (part of #277)
- Check for missing migrations in CircleCI

### Resolved issues


Fixes: #67
Fixes: #264
Fixes: #266
Fixes: #280
Fixes: #319
Fixes: #326
Fixes: #327
Fixes: #328

### Caveats

- Those changes are so substantial that I suggest to setup a new vm with the new setup and keep the old vm in case we need to rollback
- Renaming the main Django app has the inconvenient side effect that the database migrations are out of sync because the database keeps track of the migrations by app name. To resolve this issue, we have to either
  1. Manually edit the database strucure, see e.g. [this SO answer](https://stackoverflow.com/questions/8408046/how-to-change-the-name-of-a-django-app/8408131#8408131)
  2. Export the database via `vocabulary-trainer dumpdata vocgui > data.json`, then replace the old app name in the data export with `sed -i 's/vocgui/cms/g' data.json`, then import it again into the new system with `lunes-cms-cli loaddata data.json`. I did not test this solution, but if it works it might be a lot easier than option i.
